### DISCUSSION TITLE
Expected behaviour when multiple schemas are matched from schema store

### DISCUSSION THREAD

**Discussion by sapirpol** [eclipse-theia/theia#8322] [General]
### Bug Description:

When opening a `manifest.json` file in theia I get multiple schemas that are matched from schema store.
For example:

![image](https://user-images.githubusercontent.com/30637290/89557985-4f56e480-d81c-11ea-9f7a-ef4978e5b08b.png)

Here you can see auto-complete suggestions coming from both [foxx-manifest schema](https://github.com/SchemaStore/schemastore/blob/f3183484dbc0ed4504651fd6efc1a6035842823b/src/schemas/json/foxx-manifest.json#L217)  and [ui5-manifest](https://raw.githubusercontent.com/SAP/ui5-manifest/master/schema.json)

Is it a bug or a feature? 

### Steps to Reproduce:

1. Checkout theia v1.4.0.
2. Build and run theia - https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#quick-start
3. Create `webapp/manifest.json` file.
4. Try auto-complete.

**Comment by bd82** [comment]
The manifest.json name choice seems **unfortunate**. 

The SchemaStore [catalog](https://github.com/SchemaStore/schemastore/blob/master/src/api/json/catalog.json) appears to have multiple entries that can match a `manifest.json`.
- [Foxx Manifest](https://github.com/SchemaStore/schemastore/blob/d7c44777621a0780978ffd795e2f03bd066c6413/src/api/json/catalog.json#L107-L112)
- [UI5 Manifest](https://github.com/SchemaStore/schemastore/blob/d7c44777621a0780978ffd795e2f03bd066c6413/src/api/json/catalog.json#L1371-L1376)
- [Web Manifest](https://github.com/SchemaStore/schemastore/blob/d7c44777621a0780978ffd795e2f03bd066c6413/src/api/json/catalog.json#L1455-L1460)
- [Web Extensions](https://github.com/SchemaStore/schemastore/blob/d7c44777621a0780978ffd795e2f03bd066c6413/src/api/json/catalog.json#L1449-L1454)

Is there any way for the end user to pick & choose which schemas should be applied rather than always apply all?

**Reply by akosyakov** [reply]
Have you tried to use `$scheme` attribute and directly point to schema which you want to use instead?

**Reply by sapirpol** [reply]
Yes, it works.
Is there any other way to overwrite the schema to the one I want to use instead?

**Reply by akosyakov** [reply]
I would expect that as a user should be able to configure file matches and they should take a precedence over file matches contributed by extensions. If it is not it is a bug.

**Reply by bd82** [reply]
I believe that `contributes.jsonValidation` at the package.json level of extensions do **not** take precedence over
schemaStore schema matches, instead all matching schemas are applied for either source.

From an implementation perspective it seems that the **origin** of each registered fileMatch/schema pair should be "saved" and given a precedence and that **only schemas from the highest precedence** level be applied to a specific file.

I believe the precedence levels should be.
1. "$schema" prop in file.
2. user/workspace settings.json fileMatch.
3. **Identical** Precedence for these two:
   - jsonContribution package.json fileMatch. 
   - Schemas registered by Theia Extensions.
4. SchemaStore fileMatches.



@akosyakov WDYT?

**Reply by akosyakov** [reply]
Yes, it sounds great. In case of Theia there are also matches from Theia extensions, I believe they should be part of 3 or between 3 and 4. Could you convert it into the issue?

**Reply by bd82** [reply]
> Could you convert it into the issue?

I added the schemas from Theia extensions to the list.
Do you want me to open a **new issue** from the conclusions of this discussion?

**Reply by akosyakov** [reply]
yes, it would be good. I am not sure abut the solution though.

**Reply by bd82** [reply]
I've create a feature request issue here:
 https://github.com/eclipse-theia/theia/issues/8392
